Makefile

```diff
@@ -1,69 +1,83 @@
+    # Note that while ~ usually refers to $HOME, in the case where $HOME is unset,
+    # it looks up the current user's home dir, which is what we want.
+    # https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html
+    export HOME := $(shell cd ~ && pwd)
+endif
 
 ifeq ($(OS_NAME),Linux)
-	RUNTIME_ID := $(shell . /etc/os-release && echo $$ID.$$VERSION_ID)-x64
+    RUNTIME_ID := $(shell . /etc/os-release && echo $$ID.$$VERSION_ID)-x64
 else ifeq ($(OS_NAME),Darwin)
-	RUNTIME_ID := osx.10.12-x64
+    RUNTIME_ID := osx.10.12-x64
+else
+    $(error "Unknown OS_NAME: $(OS_NAME)")
 endif
 
+MSBUILD_ARGS := -c $(BUILD_CONFIGURATION) '-consoleloggerparameters:Verbosity=minimal;summary'
+
 ifneq ($(BUILD_LOG_PATH),)
-	MSBUILD_ADDITIONALARGS := $(MSBUILD_ADDITIONALARGS) /fileloggerparameters:LogFile=$(BUILD_LOG_PATH)
+    MSBUILD_ARGS += -filelogger '-fileloggerparameters:Verbosity=normal;logFile=$(BUILD_LOG_PATH)'
 endif
 
+
+BUILD_CMD := dotnet build $(MSBUILD_ARGS)
+BOOTSTRAP_CMD := dotnet publish -r $(RUNTIME_ID) $(MSBUILD_ARGS)
+
+# This gets a bit complex. There are two cases here:
+# BOOTSTRAP=false:
+#   Things proceed simply. The "all" target does not depend on the bootstrap
+#   target, so bootstrap is never built, and BootstrapBuildPath is unspecified.
+# BOOTSTRAP=true:
+#   BOOTSTRAP_DEPENDENCY is set to "bootstrap", making the "all" target depend
+#   on it, and so the bootstrap compiler gets built. Additionally,
+#   BootstrapBuildPath is specified, but *only* for the main build, *not* the
+#   bootstrap build.
 ifeq ($(BOOTSTRAP),true)
-	MSBUILD_ARGS = $(MSBUILD_ADDITIONALARGS) /p:CscToolPath=$(BOOTSTRAP_PATH)/csc /p:CscToolExe=csc /p:VbcToolPath=$(BOOTSTRAP_PATH)/vbc /p:VbcToolExe=vbc
+    BUILD_CMD += -p:BootstrapBuildPath=$(BOOTSTRAP_PATH)
+    BOOTSTRAP_DEPENDENCY := bootstrap
 else
-	MSBUILD_ARGS = $(MSBUILD_ADDITIONALARGS)
+    BOOTSTRAP_DEPENDENCY :=
 endif
 
-BUILD_CMD = dotnet build $(MSBUILD_ARGS)
-
-.PHONY: all bootstrap test toolset
+.PHONY: all bootstrap test restore
 
-all: restore
-	@export PATH="$(BINARIES_PATH)/dotnet-cli:$(PATH)" ; \
-	export HOME="$(HOME_DIR)" ; \
-	$(BUILD_CMD) CrossPlatform.sln
+all: restore $(BOOTSTRAP_DEPENDENCY)
+	@echo Building CrossPlatform.sln
+	$(BUILD_CMD) $(THIS_MAKEFILE_PATH)CrossPlatform.sln
 
 bootstrap: restore
-	export HOME="$(HOME_DIR)" ; \
-	export PATH="$(BINARIES_PATH)/dotnet-cli:$(PATH)" ; \
-	$(BUILD_CMD) src/Compilers/CSharp/CscCore && \
-	$(BUILD_CMD) src/Compilers/VisualBasic/VbcCore && \
-	mkdir -p $(BOOTSTRAP_PATH)/csc && mkdir -p $(BOOTSTRAP_PATH)/vbc && \
-	dotnet publish -c $(BUILD_CONFIGURATION) -r $(RUNTIME_ID) src/Compilers/CSharp/CscCore -o $(BOOTSTRAP_PATH)/csc && \
-	dotnet publish -c $(BUILD_CONFIGURATION) -r $(RUNTIME_ID) src/Compilers/VisualBasic/VbcCore -o $(BOOTSTRAP_PATH)/vbc
-	rm -rf Binaries/$(BUILD_CONFIGURATION)
+	@echo Building Bootstrap
+	$(BOOTSTRAP_CMD) $(SRC_PATH)/Compilers/CSharp/CscCore -o $(BOOTSTRAP_PATH)/csc
```

Comment:
Isn't `-o` a `dotnet build` parameter, while the rest are `msbuild` parameters? Maybe we should try to keep the `dotnet build` parameters first, since I don't know how `dotnet build` decides to pass parameters to msbuild.